### PR TITLE
Scene manager

### PR DESCRIPTION
**Pull request by maxdevos49**
## Summary
- Add `SceneManager` class
- Add `IScene` interface
- Add `Scene` abstract class which implements `IScene` with basic common functionality
- Added `Scenes/HelloScene.cs` to demonstrate the scene manager. Left click to change scenes.
- Added `Scenes/WorldScene.cs` to demonstrate the scene manager. Left click to change scenes.
- Tested the class `SceneManager`

## Screenshots

HelloScene:
Screenshot 2024-03-14 at 7 25 37 PM

WorldScene:
Screenshot 2024-03-14 at 7 25 39 PM

### CODE REVIEW

File: GateAdventures/Engine/Scenes/Scene.cs
Comment (danielbarnes175) [inline GateAdventures/Engine/Scenes/Scene.cs]: Personally I'd prefer if we keep these variables camel case ex. `graphicsDeviceManager` but I don't think I'm picky if you want to keep it

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (danielbarnes175) [inline GateAdventures/Engine/Scenes/SceneManager.cs:17]: What's the point of having both `_scenes` and `Scenes`?

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (danielbarnes175) [inline GateAdventures/Engine/Scenes/SceneManager.cs]: Remove this new line on L#42

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (danielbarnes175) [inline GateAdventures/Engine/Scenes/SceneManager.cs]: Remove new line L#55

File: GateAdventures/GateAdventures.cs
Comment (danielbarnes175) [inline GateAdventures/GateAdventures.cs]: If `scenes` is empty, this would set `sceneCount` to `-1`, which doesn't make any sense. We should remove the `- 1` here, and make the for loop `i < sceneCount` imo

File: GateAdventures/Scenes/HelloScene.cs
Comment (danielbarnes175) [inline GateAdventures/Scenes/HelloScene.cs]: This is the same as in `WorldScene`, I know it's just an example, but we should move this to a helper function. It'll be good practice for when we do have shared input functionality between scenes

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (danielbarnes175) [inline GateAdventures/Engine/Scenes/SceneManager.cs:41]: One of the main issues I noticed when testing the branch was that the ball would reset to the center every time we changed scenes. We don't want to reinitialize on scene reload, that should only be called on the very first load. We only should load the content.

File: GateAdventures/Scenes/HelloScene.cs
Comment (danielbarnes175) [inline GateAdventures/Scenes/HelloScene.cs]: I'm a bit nervous about each scene having it's own SpriteBatch. Seems like it defeats a lot of the point for why you would use a SpriteBatch in the first place, but I think it's probably fine for now. Just a point that we might want to discuss

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (danielbarnes175) [inline GateAdventures/Engine/Scenes/SceneManager.cs:63]: Would a `StopAll` be helpful here as well? I'm thinking something like you have the game scene, bring up a menu, then click to go the main menu, you would need to stop both the menu scene and the game scene

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (danielbarnes175) [inline GateAdventures/Engine/Scenes/SceneManager.cs]: Can we put this in a separate file?

File: GateAdventures/Engine/Scenes/Scene.cs
Comment (danielbarnes175) [reply GateAdventures/Engine/Scenes/Scene.cs]: Especially in downstream classes, i.e. HelloScene.cs, etc, it might be more clear that it's an instance variable with camel case, rather than think it's something like an imported class

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (maxdevos49) [reply GateAdventures/Engine/Scenes/SceneManager.cs:17]: This prevents external mutation of _scenes without going through the Start, Stop, or StartOverlay methods. I probably should adjust this comment to say, "Get a array of active scenes."

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (maxdevos49) [reply GateAdventures/Engine/Scenes/SceneManager.cs:63]: When you call `Start` all previously existing scenes are stopped automatically. I did not see a case where we would want to stop all scenes and not have at least one scene left. Should I still add `StopAll`?

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (maxdevos49) [reply GateAdventures/Engine/Scenes/SceneManager.cs]: Yeah I can do that

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (maxdevos49) [reply GateAdventures/Engine/Scenes/SceneManager.cs:41]: The ball resets because it is a new instance of a scene which is by design. If we need persistent state then that state probably does not belong to the scene. We can consider creating a state manager in the future which is capable of saving to a file

File: GateAdventures/Scenes/HelloScene.cs
Comment (maxdevos49) [reply GateAdventures/Scenes/HelloScene.cs]: Yes all this code is nearly identical but I only intend these classes to be short lived examples until we get farther along! I am not certain a helper function would be worth it right now

File: GateAdventures/GateAdventures.cs
Comment (maxdevos49) [reply GateAdventures/GateAdventures.cs]: That makes sense. I did not consider there ever being zero scenes

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (danielbarnes175) [reply GateAdventures/Engine/Scenes/SceneManager.cs:63]: Fair enough, I'm ok not having `StopAll` then

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (danielbarnes175) [reply GateAdventures/Engine/Scenes/SceneManager.cs:17]: That makes sense to me, basically a way for external classes to get the active scenes without being able to then change those active scenes 👍

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (danielbarnes175) [reply GateAdventures/Engine/Scenes/SceneManager.cs:41]: > The ball resets because it is a new instance of a scene which is by design

`Start()` shouldn't be a new instance. Imagine you're moving back and forth between a dungeon and the overworld, you don't want to be creating a new instance every single time. If we want a new reset instance, then we would call the constructor, and actually create a new instance, not use `Start()` as a pseudo constructor/reset.

> If we need persistent state then that state probably does not belong to the scene.

Even with a state manager, that would be a part of a scene, no? The NPCs, and objects are specific to a single scene, so that inherently means it belongs the scene. Maybe I'm misunderstanding what you mean here

> We can consider creating a state manager in the future which is capable of saving to a file

I'm not completely against this, it would be helpful to protect against crashes. Kind of like an autosave feature, however, the overhead of needing to read/write files is not insignificant, and if you did that every time you're loading and unloading a scene, if someone in the game is able to change scenes rapidly, that could lead to cases where you're reading the file before the previous close has overwritten it, which would then cause bad behavior.

File: GateAdventures/Scenes/HelloScene.cs
Comment (danielbarnes175) [reply GateAdventures/Scenes/HelloScene.cs]: Works for me

File: GateAdventures/Engine/Scenes/SceneManager.cs
Comment (maxdevos49) [reply GateAdventures/Engine/Scenes/SceneManager.cs:41]: > Start() shouldn't be a new instance. Imagine you're moving back and forth between a dungeon and the overworld, you don't want to be creating a new instance every single time. If we want a new reset instance, then we would call the constructor, and actually create a new instance, not use Start() as a pseudo constructor/reset.

The reason I went with `Start` being passed a fresh instance of a scene and instances not being reused is so the following could be done:
```
SceneManager.Start(new TileScene("map-1"));
// And then later
SceneManager.Start(new TileScene("map-2")); // map-1 scene is stopped and disposed of automatically here.
```

> Even with a state manager, that would be a part of a scene, no? The NPCs, and objects are specific to a single scene, so that inherently means it belongs the scene. Maybe I'm misunderstanding what you mean here

I mean that the scene should not contain the business logic for saving the scene. You might need to call a `Save` method in a scene but that logic does not belong to the scene.

> I'm not completely against this, it would be helpful to protect against crashes. Kind of like an autosave feature, however, the overhead of needing to read/write files is not insignificant, and if you did that every time you're loading and unloading a scene, if someone in the game is able to change scenes rapidly, that could lead to cases where you're reading the file before the previous close has overwritten it, which would then cause bad behavior.

I think this detail should be discussed later but yes I agree straight to a file would not be ok. We would likely want to have a in memory segmented cache which is only persisted to a file under certain cases(Completing a level, opening the options menu,...). If memory becomes a problem it could be modeled as a queue where the last n number of scenes visited remain in memory.

Comment (danielbarnes175) [reply GateAdventures/Engine/Scenes/SceneManager.cs:41]: Agree with the last two points. For the first one, I'm undecided. I think I just need some time to think about it, so I won't block the PR on it. To me, stopping a scene does not necessarily mean completely delete the scene, just stop drawing and updating it. Allowing us to then restart the scene later, almost like a pause functionality

File: GateAdventures/Engine/Scenes/Scene.cs
Comment (danielbarnes175) [reply GateAdventures/Engine/Scenes/Scene.cs]: I rescind this

**Review by danielbarnes175 — APPROVED** [review]
LGTM once all CI passes

File: GateAdventures/Engine/Scenes/Scene.cs
Comment (maxdevos49) [reply GateAdventures/Engine/Scenes/Scene.cs]: Camel case is reserved for method body variables or params based on C# guidelines. I added underscores to these instead

File: GateAdventures/Scenes/HelloScene.cs
Comment (maxdevos49) [reply GateAdventures/Scenes/HelloScene.cs]: I will note(I need to verify though) if we want multiple shaders in the future you then need multiple sprite batch groups. This is however is a shared instance between all scenes

Comment (maxdevos49) [reply GateAdventures/Scenes/HelloScene.cs]: We could add a second SpriteBatch which you are expected to not call `Begin` or `End` with and can be used for anything that does not require a shader.

**Pull request merged**
